go/ct/gen/transient_storage.go

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package gen
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"golang.org/x/exp/maps"
+	"pgregory.net/rand"
+
+	"github.com/Fantom-foundation/Tosca/go/ct/common"
+	"github.com/Fantom-foundation/Tosca/go/ct/st"
+)
+
+type TransientStorageGenerator struct {
+	nonZeroConstraints map[Variable]bool
+	unsatisfiable      bool
+}
+
+func NewTransientStorageGenerator() *TransientStorageGenerator {
+	return &TransientStorageGenerator{nonZeroConstraints: map[Variable]bool{}}
+}
+
+// BindToNonZero binds a location in transient storage to a non-zero value.
+func (t *TransientStorageGenerator) BindToNonZero(key Variable) {
+	if value, exits := t.nonZeroConstraints[key]; exits {
+		if !value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.nonZeroConstraints[key] = true
+	}
+}
+
+// BindToZero binds a location in transient storage to a zero value.
+func (t *TransientStorageGenerator) BindToZero(key Variable) {
+	if value, exits := t.nonZeroConstraints[key]; exits {
+		if value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.nonZeroConstraints[key] = false
+	}
+}
+
+func (t *TransientStorageGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.TransientStorage, error) {
+	if t.unsatisfiable {
+		return nil, fmt.Errorf("%w, conflicting set/unset constraints", ErrUnsatisfiable)
+	}
+
+	keysInUse := map[common.U256]bool{}
+	for variable := range t.nonZeroConstraints {
+		if pos, isBound := assignment[variable]; isBound {
+			keysInUse[pos] = true
+		}
+	}
+
+	getUnusedKey := func() common.U256 {
+		for {
+			key := common.RandU256(rnd)
+			if _, isPresent := keysInUse[key]; !isPresent {
+				keysInUse[key] = true
+				return key
+			}
+		}
+	}
+
+	transientStorage := &st.TransientStorage{}
+
+	// Process zero/nonZero constraints.
+	for variable, nonZero := range t.nonZeroConstraints {
+
+		// Check different assignments with the same key if they conflict
+		for other, otherNonZero := range t.nonZeroConstraints {
+			_, isBound := assignment[variable]
+			if assignment[variable] == assignment[other] && nonZero != otherNonZero && isBound {
```

Comment:
nit: the `isBound` check could be done before the loop; if it is not bound, the loop can be skipped.